.github/workflows/ci.yml

```diff
@@ -201,15 +283,88 @@ jobs:
 
       - name: Build end-user environment
         run: |
-          ./dev_scripts/env.py --distro fedora --version ${{ matrix.version }} \
+          ./dev_scripts/env.py --distro ${{ matrix.distro }} \
+              --version ${{ matrix.version }} \
               build --download-pyside6
 
       - name: Run a test command
         run: |
-          ./dev_scripts/env.py --distro fedora --version ${{ matrix.version }} \
+          ./dev_scripts/env.py --distro ${{ matrix.distro }} --version ${{ matrix.version }} \
               run dangerzone-cli dangerzone/tests/test_docs/sample-pdf.pdf
 
       - name: Check that the Dangerzone GUI imports work
         run: |
-          ./dev_scripts/env.py --distro fedora --version ${{ matrix.version }} \
+          ./dev_scripts/env.py --distro ${{ matrix.distro }} --version ${{ matrix.version }} \
               run dangerzone --help
+
+  run-tests:
+    name: "run tests (${{ matrix.distro }} ${{ matrix.version }})"
+    runs-on: ubuntu-latest
+    needs: build-container-image
+    strategy:
+      matrix:
+        include:
+          - distro: ubuntu
+            version: "20.04"
+          - distro: ubuntu
+            version: "22.04"
+          - distro: ubuntu
+            version: "23.10"
+          - distro: ubuntu
+            version: "24.04"
+          - distro: debian
+            version: bullseye
+          - distro: debian
+            version: bookworm
+          - distro: debian
+            version: trixie
+          - distro: fedora
+            version: "39"
+          - distro: fedora
+            version: "40"
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.10"
+
+      - name: Login to GHCR
+        run: |
+          echo ${{ github.token }} | podman login ghcr.io -u USERNAME --password-stdin
+
+      - name: Set modified version variable
+        run: |
+          if [ "${{ matrix.version }}" = "bullseye" ]; then
+            echo "modified_version=${{ matrix.version }}-backports" >> $GITHUB_ENV
+          else
+            echo "modified_version=${{ matrix.version }}" >> $GITHUB_ENV
+          fi
+
+      - name: Install dev env from GHCR
+        run: |
+          podman pull \
+            ${{ env.IMAGE_REGISTRY }}/dangerzone/build/${{ matrix.distro }}:${{ env.modified_version }}
```

Comment:
You've probably have alluded to this already, but what happens here if the `env.py` script changes, and creates the dev environments slightly different for a PR?

I think we should hash the resources that shape our dev images (much as you have done already for the container image), and make this hash past of the container image tag. To make things simpler, I'd suggest adding a `--sync` flag in our `env.py` script, which will attempt to pull the image, build it if not already built, and then push it to GHCR.